A58-client-side-weighted-round-robin-lb-policy.md

```diff
@@ -0,0 +1,251 @@
+
+When `oob_reporting_period` changes, the client implementation starts a new OOB load reporting stream as outlined in [gRFC A51].
+
+Changes to `weight_update_period`, `blackout_period`, and `weight_expiration_period` take effective with a new picker that we always create upon receiving a resolver update.
+
+In all languages, `weighted_round_robin` policy de-duplicates redundant addresses in resolver updates. This behavior is different from `round_robin` where it uses the number of representations in the list as a resolver supplied weight.
+
+#### Managing Subchannels
+
+The behavior remains the same as in `round_robin` in all languages.
+
+In C/C++, the control plane creates a new pending subchannel list upon getting a new address list and starts watching the connectivity state of its subchannels. Only subchannels in the `READY` state are included. The pending list is immediately promoted to the current list if it’s an initial update or the new list is empty (i.e. no subchannel). Otherwise the pending list is later promoted to the current list if the current list has none `READY`, or the pending list has at least one `READY`, or all subchannels in the pending list are `TRANSIENT_FAILURE`. The rest of the process is triggered upon the connectivity state of a subchannel changes in either the pending or current list. When the connectivity state of a subchannel changes and it is in the pending subchannel list, it checks whether to promote the pending list to current. If it’s in the current list it updates the aggregate connectivity state and creates a new picker. The kind of picker created depends on the aggregate state: an active Picker (w/ a scheduler) if `READY`, a QueuePicker if `CONNECTING`, or a TransientFailurePicker otherwise.
+
+In Java, the current subchannel list is updated directly as in `round_robin`. Upon getting an updated address list, it adds new subchannels to the subchannel list, starts watching the connectivity state of it, and requests connection. It also deletes removed subchannels from the subchannel list and shuts them down. It then updates the aggregate connectivity state, and if there is at least one `READY` subchannel, creates a new picker. The kind of picker created depends on the aggregate state: an active Picker (with a scheduler) if `READY`, EmptyPicker otherwise. When the connectivity state of a subchannel changes, it triggers updating the aggregate connectivity state.
+
+In Go, `baseBalancer` handles the address update as in `round_robin`. When the address set changes, the `baseBuilder` calls the picker builder of `weighted_round_robin` if the aggregate connectivity state is `READY`. When the connectivity state of a subchannel changes, it triggers the same process as when the address set changes.
+
+In all languages, when a subchannel state is updated to `IDLE`, the control plane immediately requests a connection and treats it as if it is in `CONNECTING` for the aggregation purpose.
+
+### Subchannel Weights
```

Comment:
Please add the following paragraph to this section:

"""
In Java and Go, the weight will be stored in a wrapped subchannel.  However, in C++, that approach won't work, because we create a new subchannel object for each address whenever we get a new address list, and we don't want to lose existing weight information when that happens.  So instead, we store the weights in a map keyed by address, and each subchannel list will take its own ref to the entry in the map for each subchannel in the list.
"""

Reply:
Done.